### PR TITLE
Fix the issue in Windows kube-proxy when processing unqualified name. This is for DNS client such as ping or iwr that validate name in response and original question.

### LINKED ISSUE #42605: DNS resolving does not work on windows nodes
BUG REPORT
## Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.5", GitCommit:"5a0a696437ad35c133c0c8493f7e9d22b0f9b81b", GitTreeState:"clean", BuildDate:"2016-10-29T01:38:40Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"windows/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.3", GitCommit:"029c3a408176b55c30846f0faedf56aae5992e9b", GitTreeState:"clean", BuildDate:"2017-02-15T06:34:56Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}

## Environment**:
Azure ACS with windows agents
**What happened**:
unable to connect from one deployed service to other service based on the documented kubernetes dns name of the deployed service

## repro steps:
Created two services using the following commands:
kubectl create deployment iis --image=microsoft/iis
kubectl expose iis --port:80 --type=LoadBalancer
 service is now available on external network
now I want from in this container to connect to a database using the dns name db.default.svc.cluster.local
so created new service:
kubectl create deployment db --image=marcelv/iisexpress
and expose this internal in the cluster
kubectl create service clusterip db --tcp=1433:1433

now I execute a command in the iis container to see if I can connect to the new db service:
kubectl exec <nameofthepod> cmd.exe
inside container I execute:
nslookup
db.default.svc.cluster.local
this results the ip address I also see when I call the kubectl get services command
but when I want to connect to the dns name on port 1433 it wont resolve the ip adres
if you do
ping db.default.svc.cluster.local
Ping request could not find host db.default.cluster.local. Please check the name and try again.

so with a direct dns query it works, but the implicit name resolving on windows does not work.
I tried to resolve it using 
ipconfig /registerdns
no avail
looking at the DNS cash does not show any entries

So it seems DNS is not fully functional inside the cluster.

## What you expected to happen**:
I expect to be able to connect to the network via DNS  from the pod
I also deployed a .net application that tried to resolve the address
but those fal with a wind32 error network path not found

### PR DESCRIPTION
**Pull request by JiangtianLi**
**What this PR does / why we need it**:
This PR is an additional fix to #41618 and [the corresponding commit](https://github.com/kubernetes/kubernetes/commit/b9dfb69dd7679ae31bad1fb4980561fea22f8ed4). The DNS client such as nslookup does not validate name matching in response and original question. That works fine when we append DNS suffix to unqualified name in DNS query in Windows kube-proxy. However, for DNS client such as ping or Invoke-WebRequest that validates name in response and original question, the issue arises and the DNS query fails although the received DNS response has no error. 

This PR fixes the additional issue by restoring the original question name in DNS response. Further, this PR refactors DNS message routines by using miekg's DNS library.

This PR affects the Windows kube-proxy only.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #42605 

**Special notes for your reviewer**:

**Release note**:

```release-note
Fix DNS suffix search list support in Windows kube-proxy.
```

### CODE REVIEW

**Comment by spiffxp** [comment]
@k8s-bot ok to test

**Comment by JiangtianLi** [comment]
Compared to another succeeded tests (build 23822) right after `docker run`:
```
W0508 13:47:05.063] Unable to find image 'gcr.io/k8s-testimages/kubekins-e2e:v20170324-331e93f3' locally
I0508 13:47:05.163] Starting...
W0508 13:47:05.512] v20170324-331e93f3: Pulling from k8s-testimages/kubekins-e2e
W0508 13:47:05.513] 966c0c9c92c0: Pulling fs layer
```
This failed test (build 23821) log has instead:
```
I0508 13:44:57.211] Starting...
W0508 13:44:57.754] Error response from daemon: could not find image: no such id: c99be84a44929eabbb1ef65012d2e118b3d1ec3733375e53d3d3bc7e8245a531
W0508 13:44:57.755] + rc=1
W0508 13:44:57.756] + echo 'Exiting with code: 1'
```

It seems gcr.io/k8s-testimages/kubekins-e2e:v20170324-331e93f3 was found locally but not found by image id? Any idea where to investigate further?

File: pkg/proxy/winuserspace/proxysocket.go [hunk 1]
```diff
@@ -436,44 +252,85 @@ type dnsClientCache struct {
...
+	origName := msg.Question[0].Name
 	if dnsSuffix != "" {
-		msg.question[0].qName.name += "." + dnsSuffix
+		msg.Question[0].Name += dnsSuffix + "."
+	}
+	mbuf, err := msg.PackBuffer(buffer)
+	msg.Question[0].Name = origName
+
+	if err != nil {
+		glog.Warning("Unable to pack DNS packet. Error is: %v", err)
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxysocket.go]: hrm, I think we should change the return to (int, err)

File: pkg/proxy/winuserspace/proxysocket.go [hunk 2]
```diff
@@ -436,44 +252,85 @@ type dnsClientCache struct {
...
+	}
+
+	msg.Question[0].Name = origName
+	if len(msg.Answer) > 0 {
+		msg.Answer[0].Header().Name = origName
+	}
+	mbuf, err := msg.PackBuffer(buffer)
+
+	if err != nil {
+		glog.Warning("Unable to pack DNS packet. Error is: %v", err)
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxysocket.go]: as above, return (int, error)

File: pkg/proxy/winuserspace/proxysocket.go [hunk 3]
```diff
@@ -527,98 +393,96 @@ func processUnpackedDnsResponsePacket(svrConn net.Conn, dnsClients *dnsClientCac
...
 }
 
-func processDnsQueryPacket(dnsClients *dnsClientCache, cliAddr net.Addr, buffer []byte, length int, dnsSearch []string) int {
-	msg := &dnsMsg{}
-	if !msg.unpackDnsMsg(buffer[:length]) {
-		glog.Warning("Unable to unpack DNS packet.")
+func processDNSQueryPacket(dnsClients *dnsClientCache, cliAddr net.Addr, buffer []byte, length int, dnsSearch []string) int {
+	msg := &dns.Msg{}
+	if err := msg.Unpack(buffer[:length]); err != nil {
+		glog.Warning("Unable to unpack DNS packet. Error is: %v", err)
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxysocket.go]: here too.

File: pkg/proxy/winuserspace/proxysocket.go [hunk 4]
```diff
@@ -527,98 +393,96 @@ func processUnpackedDnsResponsePacket(svrConn net.Conn, dnsClients *dnsClientCac
...
-func processDnsResponsePacket(svrConn net.Conn, dnsClients *dnsClientCache, cliAddr net.Addr, buffer []byte, length int, dnsSearch []string) bool {
+func processDNSResponsePacket(svrConn net.Conn, dnsClients *dnsClientCache, cliAddr net.Addr, buffer []byte, length int, dnsSearch []string) (bool, int) {
 	var drop bool
-	msg := &dnsMsg{}
-	if !msg.unpackDnsMsg(buffer[:length]) {
-		glog.Warning("Unable to unpack DNS packet.")
-		return drop
+	msg := &dns.Msg{}
+	if err := msg.Unpack(buffer[:length]); err != nil {
+		glog.Warning("Unable to unpack DNS packet. Error is: %v", err)
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxysocket.go]: here too...

**Comment by brendandburns** [comment]
@k8s-bot kops aws e2e test this

Some comments on returning errors for error cases.

**Comment by JiangtianLi** [comment]
Hmm... I have rebased and pushed in an incorrect way. Will correct this.

**Comment by JiangtianLi** [comment]
Sorry for wasting resources. Closing this and created #45642.

File: pkg/proxy/winuserspace/proxysocket.go (hunk 1 above)
Comment (JiangtianLi) [inline pkg/proxy/winuserspace/proxysocket.go]: Sure. I have made the change. However, I didn't push correctly so this branch and PR appears to be correct. I have moved to #45642 to address your comment in a new commit. The commits can be squashed later and for now it is easier to review the changes. Sorry this is a bit cumbersome.

**Comment by JiangtianLi** [comment]
@brendandburns Due to my bad rebase and push as I commented before, I've closed this PR and created #45642 to address your comment. There is conversation started in the new PR. Sorry for the inconvenience.

**Pull request closed without merging**